File: 8vim/src/main/java/inc/flide/vim8/keyboardhelpers/KeyboardDataYamlParser.java
package inc.flide.vim8.keyboardhelpers;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.MapperFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.fasterxml.jackson.databind.module.SimpleModule;
import com.fasterxml.jackson.dataformat.yaml.YAMLMapper;
import com.networknt.schema.JsonSchema;
import com.networknt.schema.JsonSchemaException;
import com.networknt.schema.JsonSchemaFactory;
import com.networknt.schema.SpecVersionDetector;
import com.networknt.schema.ValidationMessage;

import inc.flide.vim8.structures.Constants;
import inc.flide.vim8.structures.KeyboardAction;
import inc.flide.vim8.structures.KeyboardData;
import inc.flide.vim8.structures.Quadrant;
import inc.flide.vim8.structures.exceptions.InvalidYamlException;
import inc.flide.vim8.structures.exceptions.YamlException;
import inc.flide.vim8.structures.exceptions.YamlParsingException;
import inc.flide.vim8.structures.yaml.Action;
import inc.flide.vim8.structures.yaml.ExtraLayer;
import inc.flide.vim8.structures.yaml.Flags;
import inc.flide.vim8.structures.yaml.Layer;
import inc.flide.vim8.structures.yaml.Layout;
import inc.flide.vim8.structures.yaml.Part;
import inc.flide.vim8.utils.MovementSequenceHelper;
import java.io.IOException;
import java.io.InputStream;
import java.text.MessageFormat;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;
import org.apache.commons.lang3.tuple.Pair;

public class KeyboardDataYamlParser {
    private static final SimpleModule module =
            new SimpleModule().addDeserializer(Flags.class, new Flags.FlagsDeserializer());
    private static final ObjectMapper mapper =
            YAMLMapper.builder().enable(MapperFeature.ACCEPT_CASE_INSENSITIVE_ENUMS)
                    .propertyNamingStrategy(PropertyNamingStrategies.SNAKE_CASE)
                    .addModule(module).build();
    private static final JsonSchema schema;

    static {
        try (InputStream schemaInputStream = KeyboardDataYamlParser.class.getResourceAsStream("/schema.json")) {
            JsonNode schemaJson = mapper.readTree(schemaInputStream);
            JsonSchemaFactory factory =
                    JsonSchemaFactory.builder(
                                    JsonSchemaFactory.getInstance(SpecVersionDetector.detect(schemaJson)))
                            .objectMapper(mapper)
                            .build();
            schema = factory.getSchema(schemaJson);
        } catch (IOException e) {
            throw new YamlParsingException(e);
        } catch (JsonSchemaException e) {
            throw new InvalidYamlException(e.getMessage());
        }
    }

    public static KeyboardData readKeyboardData(InputStream inputStream) throws YamlException {
        try {
            JsonNode node = mapper.readTree(inputStream);
            validateYaml(node);
            Layout layout = mapper.treeToValue(node, Layout.class);
            KeyboardData keyboardData = new KeyboardData();
            keyboardData.setInfo(layout.info);

            if (!layout.layers.hidden.isEmpty()) {
                addKeyboardActions(keyboardData, layout.layers.hidden);
            }

            Layer defaultLayer = layout.layers.defaultLayer;
            Map<ExtraLayer, Layer> extraLayers = layout.layers.extraLayers;

            if (!extraLayers.isEmpty() && defaultLayer == null) {
                return keyboardData;
            }

            // Find the longest action chain to suze the keyboard
            if (defaultLayer != null) {
                for (Map.Entry<Integer, Part> sectorEntry : defaultLayer.sectors.entrySet()) {
                    int sector = sectorEntry.getKey();
                    keyboardData.sectors = Math.max(keyboardData.sectors, sector);
                    for (Map.Entry<Integer, List<Action>> partEntry : sectorEntry.getValue().parts.entrySet()) {
                        keyboardData.layoutPositions = Math.max(keyboardData.layoutPositions, partEntry.getValue().size());
                    }
                }
            }
            for (Map.Entry<ExtraLayer, Layer> extraLayerMapEntry : extraLayers.entrySet()) {
                for (Map.Entry<Integer, Part> sectorEntry : extraLayerMapEntry.getValue().sectors.entrySet()) {
                    int sector = sectorEntry.getKey();
                    keyboardData.sectors = Math.max(keyboardData.sectors, sector);
                    for (Map.Entry<Integer, List<Action>> partEntry : sectorEntry.getValue().parts.entrySet()) {
                        keyboardData.layoutPositions = Math.max(keyboardData.layoutPositions, partEntry.getValue().size());
                    }
                }
            }

            if (defaultLayer != null) {
                addLayer(keyboardData, Constants.DEFAULT_LAYER, defaultLayer);
            }

            for (Map.Entry<ExtraLayer, Layer> extraLayerMapEntry : extraLayers.entrySet()) {
                int layer = extraLayerMapEntry.getKey().ordinal() + 2;
                addLayer(keyboardData, layer, extraLayerMapEntry.getValue());
            }
            return keyboardData;
        } catch (IOException exception) {
            throw new YamlParsingException(exception);
        }
    }

    private static void validateYaml(JsonNode node) {
        Set<ValidationMessage> result = schema.validate(node);
        if (!result.isEmpty()) {
            Set<ValidationMessage> errors = new HashSet<>();
            for (ValidationMessage error : result) {
                if (error.getMessage().startsWith("$")) {
                    errors.add(error);
                } else {
                    errors.add(new ValidationMessage.Builder()
                            .type(error.getType())
                            .code(error.getCode())
                            .path(error.getPath())
                            .details(error.getDetails())
                            .arguments(error.getMessage())
                            .format(new MessageFormat("{0}: {1}"))
                            .build());
                }
            }
            throw new InvalidYamlException(errors);
        }
    }

    private static void addLayer(KeyboardData keyboardData, int layer, Layer layerData) {
        StringBuilder lowerCaseCharacters = new StringBuilder();
        StringBuilder upperCaseCharacters = new StringBuilder();
        Pair<StringBuilder, StringBuilder> characterSets =
                Pair.of(lowerCaseCharacters, upperCaseCharacters);

        for (Map.Entry<Integer, Part> sectorEntry : layerData.sectors.entrySet()) {
            int sector = sectorEntry.getKey();

            for (Map.Entry<Integer, List<Action>> partEntry : sectorEntry.getValue().parts.entrySet()) {
                int part = partEntry.getKey();
                Quadrant quadrant = new Quadrant(sector, part);
                addKeyboardActions(keyboardData, layer, quadrant, partEntry.getValue(), characterSets);
            }
        }

        keyboardData.setLowerCaseCharacters(String.valueOf(lowerCaseCharacters), layer);
        keyboardData.setUpperCaseCharacters(String.valueOf(upperCaseCharacters), layer);
    }

    private static void addKeyboardActions(KeyboardData keyboardData, List<Action> actions) {
        for (Action action : actions) {
            List<Integer> movementSequence = action.movementSequence;

            if (movementSequence.isEmpty()) {
                continue;
            }

            KeyboardAction actionMap =
                    new KeyboardAction(action.actionType, action.lowerCase, action.upperCase,
                            action.getKeyCode(), action.flags.getValue(),
                            Constants.HIDDEN_LAYER);
            keyboardData.addActionMap(movementSequence, actionMap);
        }
    }

    private static int getCharacterSetSize(KeyboardData keyboardData) {
        return keyboardData.sectors * 2 * keyboardData.layoutPositions;
    }

    private static void addKeyboardActions(KeyboardData keyboardData, int layer, Quadrant quadrant,
                                           List<Action> actions,
                                           Pair<StringBuilder, StringBuilder> characterSets) {
        int actionsSize = Math.min(actions.size(), 4);

        for (int i = 0; i < actionsSize; i++) {
            Action action = actions.get(i);
            if (action == null || action.isEmpty()) {
                continue;
            }

            int characterPosition = i;

            List<Integer> movementSequence = action.movementSequence;

            if (movementSequence.isEmpty()) {
                movementSequence =
                        MovementSequenceHelper.computeMovementSequence(layer, quadrant, characterPosition, keyboardData);
            }

            int characterSetIndex = quadrant.getCharacterIndexInString(characterPosition, keyboardData);

            if (!action.lowerCase.isEmpty()) {
                if (characterSets.getLeft().length() == 0) {
                    characterSets.getLeft().setLength(getCharacterSetSize(keyboardData));
                }
                characterSets.getLeft().setCharAt(characterSetIndex, action.lowerCase.charAt(0));

                if (action.upperCase.isEmpty()) {
                    action.upperCase = action.lowerCase.toUpperCase(Locale.ROOT);
                }
            }

            if (!action.upperCase.isEmpty()) {
                if (characterSets.getRight().length() == 0) {
                    characterSets.getRight().setLength(getCharacterSetSize(keyboardData));
                }
                characterSets.getRight().setCharAt(characterSetIndex, action.upperCase.charAt(0));
            }

            KeyboardAction actionMap =
                    new KeyboardAction(action.actionType, action.lowerCase, action.upperCase,
                            action.getKeyCode(), action.flags.getValue(),
                            layer);

            keyboardData.addActionMap(movementSequence, actionMap);
            if (layer > Constants.DEFAULT_LAYER && action.movementSequence.isEmpty()) {
                keyboardData.addActionMap(
                        MovementSequenceHelper.computeQuickMovementSequence(layer, quadrant, characterPosition, keyboardData),
                        actionMap);
            }
        }
    }
}

File: 8vim/src/main/kotlin/inc/flide/vim8/ime/KeyboardDataYamlParser.kt
package inc.flide.vim8.ime

import arrow.core.Either
import arrow.core.fold
import arrow.core.left
import arrow.core.raise.catch
import arrow.core.right
import arrow.core.toMap
import arrow.integrations.jackson.module.registerArrowModule
import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.MapperFeature
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.PropertyNamingStrategies
import com.fasterxml.jackson.databind.module.SimpleModule
import com.fasterxml.jackson.dataformat.yaml.YAMLMapper
import com.fasterxml.jackson.module.kotlin.addDeserializer
import com.fasterxml.jackson.module.kotlin.convertValue
import com.fasterxml.jackson.module.kotlin.registerKotlinModule
import com.networknt.schema.JsonSchema
import com.networknt.schema.JsonSchemaFactory
import com.networknt.schema.SpecVersionDetector
import com.networknt.schema.ValidationMessage
import inc.flide.vim8.models.CHARACTER_SET_SIZE
import inc.flide.vim8.models.CharacterPosition
import inc.flide.vim8.models.FingerPosition
import inc.flide.vim8.models.KeyboardAction
import inc.flide.vim8.models.KeyboardData
import inc.flide.vim8.models.LayerLevel
import inc.flide.vim8.models.MovementSequence
import inc.flide.vim8.models.Quadrant
import inc.flide.vim8.models.addAllToActionMap
import inc.flide.vim8.models.characterIndexInString
import inc.flide.vim8.models.error.ExceptionWrapperError
import inc.flide.vim8.models.error.InvalidLayoutError
import inc.flide.vim8.models.error.LayoutError
import inc.flide.vim8.models.error.validationMessages
import inc.flide.vim8.models.setLowerCaseCharacters
import inc.flide.vim8.models.setUpperCaseCharacters
import inc.flide.vim8.models.yaml.Action
import inc.flide.vim8.models.yaml.Flags
import inc.flide.vim8.models.yaml.Flags.FlagsDeserializer
import inc.flide.vim8.models.yaml.Layer
import inc.flide.vim8.models.yaml.Layout
import inc.flide.vim8.models.yaml.isEmpty
import inc.flide.vim8.models.yaml.keyCode
import inc.flide.vim8.models.yaml.toLayerLevel
import java.io.IOException
import java.io.InputStream
import java.text.MessageFormat

object KeyboardDataYamlParser {
    private val module = SimpleModule(FlagsDeserializer::class.qualifiedName)
        .addDeserializer(
            Flags::class,
            FlagsDeserializer()
        )
    private val mapper: ObjectMapper =
        YAMLMapper.builder().enable(MapperFeature.ACCEPT_CASE_INSENSITIVE_ENUMS)
            .propertyNamingStrategy(PropertyNamingStrategies.SNAKE_CASE)
            .addModule(module)
            .build()
            .registerKotlinModule()
            .registerArrowModule()
    private var schema: JsonSchema

    init {
        KeyboardDataYamlParser::class.java.getResourceAsStream("/schema.json")
            .use { schemaInputStream ->
                val schemaJson = mapper.readTree(schemaInputStream)
                val factory = JsonSchemaFactory.builder(
                    JsonSchemaFactory.getInstance(SpecVersionDetector.detect(schemaJson))
                )
                    .objectMapper(mapper)
                    .build()
                schema = factory.getSchema(schemaJson)
            }
    }

    @JvmStatic
    fun readKeyboardData(inputStream: InputStream?): Either<LayoutError, KeyboardData> {
        return catch({
            validateYaml(mapper.readTree(inputStream))
                .map { layout ->
                    val keyboardData = KeyboardData(
                        info = layout.info,
                        actionMap = getActionMap(layout.layers.hidden)
                    )

                    val layersToAdd = layout.layers.defaultLayer
                        .map { LayerLevel.FIRST to it }
                        .toMap() + layout.layers.extraLayers
                        .mapKeys { it.key.toLayerLevel() }

                    layersToAdd
                        .fold(keyboardData) { acc, (layerId, layer) ->
                            addLayer(acc, layerId, layer)
                        }
                }
        }) { exception: IOException ->
            ExceptionWrapperError(exception).left()
        }
    }

    private fun validateYaml(node: JsonNode): Either<LayoutError, Layout> {
        return schema.validate(node)
            .fold(InvalidLayoutError(emptySet())) { acc, error ->
                val newError =
                    if (error.message.startsWith('$')) {
                        error
                    } else {
                        ValidationMessage.Builder()
                            .type(error.type)
                            .code(error.code)
                            .path(error.path)
                            .details(error.details)
                            .arguments(error.message)
                            .format(MessageFormat("{0}: {1}"))
                            .build()
                    }
                InvalidLayoutError.validationMessages.modify(acc) { it + newError }
            }.let {
                if (it.validationMessages.isNotEmpty()) {
                    it.left()
                } else {
                    mapper.convertValue<Layout>(node).right()
                }
            }
    }

    private fun addLayer(
        keyboardData: KeyboardData,
        layer: LayerLevel,
        layerData: Layer
    ): KeyboardData {
        val lowerCaseCharacters = StringBuilder()
        val upperCaseCharacters = StringBuilder()
        val characterSets = (lowerCaseCharacters to upperCaseCharacters)
        return layerData.sectors.fold(keyboardData) { acc, (sector, value) ->
            value.parts.fold(acc) { acc1, (part, actions) ->
                acc1.addAllToActionMap(
                    getActionMap(
                        layer = layer,
                        quadrant = Quadrant(sector, part),
                        actions = actions,
                        characterSets = characterSets
                    )
                )
            }
        }
            .setLowerCaseCharacters(lowerCaseCharacters.toString(), layer)
            .setUpperCaseCharacters(upperCaseCharacters.toString(), layer)
    }

    private fun getActionMap(actions: List<Action>): Map<MovementSequence, KeyboardAction> {
        return actions
            .filterNot { it.movementSequence.isEmpty() }
            .associateBy({ it.movementSequence }, {
                KeyboardAction(
                    it.actionType,
                    it.lowerCase,
                    it.upperCase,
                    it.keyCode(),
                    it.flags.value,
                    LayerLevel.HIDDEN
                )
            })
    }

    private fun getActionMap(
        layer: LayerLevel,
        quadrant: Quadrant,
        actions: List<Action?>,
        characterSets: Pair<StringBuilder, StringBuilder>
    ): Map<MovementSequence, KeyboardAction> {
        return actions
            .take(4)
            .withIndex()
            .filterNot { it.value.isEmpty() }
            .fold(emptyMap()) { acc, (i, action) ->
                val characterPosition = CharacterPosition.values()[i]
                var movementSequence: MovementSequence = action!!.movementSequence
                if (movementSequence.isEmpty()) {
                    movementSequence = FingerPosition.computeMovementSequence(
                        layer,
                        quadrant,
                        characterPosition
                    )
                }
                val characterSetIndex: Int =
                    quadrant.characterIndexInString(characterPosition)
                val updatedAction = if (action.lowerCase.isNotEmpty()) {
                    if (characterSets.first.isEmpty()) {
                        characterSets.first.setLength(CHARACTER_SET_SIZE)
                    }
                    characterSets.first.setCharAt(characterSetIndex, action.lowerCase[0])
                    if (action.upperCase.isEmpty()) {
                        action.copy(upperCase = action.lowerCase.uppercase())
                    } else {
                        action
                    }
                } else {
                    action
                }
                if (updatedAction.upperCase.isNotEmpty()) {
                    if (characterSets.second.isEmpty()) {
                        characterSets.second.setLength(CHARACTER_SET_SIZE)
                    }
                    characterSets.second.setCharAt(
                        characterSetIndex,
                        updatedAction.upperCase[0]
                    )
                }
                val keyboardAction = KeyboardAction(
                    updatedAction.actionType,
                    updatedAction.lowerCase,
                    updatedAction.upperCase,
                    updatedAction.keyCode(),
                    updatedAction.flags.value,
                    layer
                )
                val baseActionMap = movementSequence to keyboardAction
                val actionMap = when {
                    layer != LayerLevel.FIRST && action.movementSequence.isEmpty() -> mapOf(
                        baseActionMap,
                        FingerPosition.computeQuickMovementSequence(
                            layer,
                            quadrant,
                            characterPosition
                        ) to keyboardAction
                    )

                    else -> mapOf(baseActionMap)
                }

                acc + actionMap
            }
    }
}
